### PR TITLE
Parse ellipses in Java keyword annotations

### PR DESCRIPTION
**Pull request by nbrahms**
Allows the user to match a subset of keywords in a keyword annotation.

Fixes returntocorp/semgrep#522

### CODE REVIEW

File: lang_java/parsing/ast_java.ml [hunk 1]
```diff
@@ -118,7 +118,8 @@ type modifier =
 
  and annotation_element =
    | AnnotArgValue of element_value
-   | AnnotArgPairInit of annotation_pair list
+   (* Optional tok for ellipsis in sgrep mode *)
+   | AnnotArgPairInit of annotation_pair list * tok option
```
[the remark this answers is not in the document]
Comment (colleend) [reply lang_java/parsing/ast_java.ml]: done!

**Comment by aryx** [comment]
Hmm actually this will not allow the pattern grayson originally wrote:
@RequestMapping(..., method=$M, ...)
public $RETURN $METHOD(...) { ... }

This is because we allow ... at the beginning but not the end.
We also probably want to allow also @RequestMapping(...).

**Comment by aryx** [comment]
There's different want to solve that.
We could add a new AnnotArgEllipsis in annotation_element and make annotation_pair a variant with
  AnnotPair of ident * element_value
| AnnotPairEllipsis of tok
and then extend the grammar accordingly.

**Review by aryx — CHANGES_REQUESTED** [review]
This will not support the semgrep pattern of grayson:
@RequestMapping(..., method=$M, ...)
public $RETURN $METHOD(...) { ... }

**Comment by colleend** [comment]
> There's different want to solve that.
> We could add a new AnnotArgEllipsis in annotation_element and make annotation_pair a variant with
> AnnotPair of ident * element_value
> | AnnotPairEllipsis of tok
> and then extend the grammar accordingly.

sorry, why is it necessary to extend annotation_pair with AnnotPairEllipsis of tok? Is this so we can  support the case where there's ellipses on both sides?

**Comment by aryx** [comment]
yes. Grayson semgrep.live example was using ... on both sides, and we support ... on both sides
in arguments, parameters, classes, etc. so there's no reason to not allow it in annotation arguments.

**Comment by colleend** [comment]
I changed annotation_pair to a variant, but had to delete expr1's variant of "..." because this led to a reduce-reduce conflict :o 
@aryx if there's anything I can do to make the code cleaner, please feel free to tell me! 

Not sure if this is the correct direction, but did my best :)

**Review by aryx — APPROVED** [review]
Lookd good! I think it should work like that.

**Comment by minusworld** [comment]
YESSSS

**Pull request merged**